plugins/grpc-ecosystem/gateway/v2.15.2/separate_pkg_additional_imports.patch

```diff
@@ -0,0 +1,291 @@
+diff --git a/internal/descriptor/registry.go b/internal/descriptor/registry.go
+index 9cd14cc7..fe6c7199 100644
+--- a/internal/descriptor/registry.go
++++ b/internal/descriptor/registry.go
+@@ -146,6 +146,13 @@ type Registry struct {
+
+ 	// allowPatchFeature determines whether to use PATCH feature involving update masks (using google.protobuf.FieldMask).
+ 	allowPatchFeature bool
++
++	// separatePackage determines whether to output the generated code into a separate package.
++	separatePackage bool
++
++	// additionalImports is a list of additional imports to be added to the generated code.
++	// N.B. additional imports is not a flag option
++	additionalImports []string
+ }
+
+ type repeatedFieldSeparator struct {
+@@ -236,7 +243,9 @@ func (r *Registry) loadFile(filePath string, file *protogen.File) {
+ 	if r.standalone {
+ 		pkg.Alias = "ext" + cases.Title(language.AmericanEnglish).String(pkg.Name)
+ 	}
+-
++	if r.separatePackage {
++		pkg.Name += "gateway"
++	}
+ 	if err := r.ReserveGoPackageAlias(pkg.Name, pkg.Path); err != nil {
+ 		for i := 0; ; i++ {
+ 			alias := fmt.Sprintf("%s_%d", pkg.Name, i)
+@@ -811,3 +820,23 @@ func (r *Registry) SetAllowPatchFeature(allow bool) {
+ func (r *Registry) GetAllowPatchFeature() bool {
+ 	return r.allowPatchFeature
+ }
++
++// SetSeparatePackage sets separatePackage
++func (r *Registry) SetSeparatePackage(use bool) {
++	r.separatePackage = use
++}
++
++// GetSeparatePackage returns separatePackage
++func (r *Registry) GetSeparatePackage() bool {
++	return r.separatePackage
++}
++
++// SetAdditionalImports sets additionalImports
++func (r *Registry) SetAdditionalImports(imports []string) {
++	r.additionalImports = imports
++}
++
++// GetAdditionalImports returns additionalImports
++func (r *Registry) GetAdditionalImports() []string {
++	return r.additionalImports
++}
+diff --git a/internal/descriptor/services.go b/internal/descriptor/services.go
+index 8936a782..1fd4d8dc 100644
+--- a/internal/descriptor/services.go
++++ b/internal/descriptor/services.go
+@@ -3,6 +3,7 @@ package descriptor
+ import (
+ 	"errors"
+ 	"fmt"
++	"path/filepath"
+ 	"strings"
+
+ 	"github.com/golang/glog"
+@@ -25,6 +26,31 @@ func (r *Registry) loadServices(file *File) error {
+ 			ServiceDescriptorProto: sd,
+ 			ForcePrefixedName:      r.standalone,
+ 		}
++		if r.separatePackage {
++			// when generating a separate package for the gateway, we need to generate an import statement
++			// for the gRPC stubs that are no longer in the same package. This is done by adding the grpc
++			// package to the additionalImports list. In order to prepare a valid import statement, we'll replace
++			// the source package name, something like: ../pet/v1/v1petgateway with ../pet/v1/v1petgrpc
++			const (
++				baseTypePackageName = "protocolbuffers"
++				grpcPackageName     = "grpc"
++			)
++			packageName := strings.ReplaceAll(svc.File.GoPkg.Name, "gateway", "") + grpcPackageName
```

Comment:
I'm worried about doing `ReplaceAll` here - don't we just want to trim the `gateway` suffix here and then append grpc?